specs/TryUpdateRuntime.md

```diff
@@ -0,0 +1,229 @@
+  /// See `UpdateError` property of UpdateRuntimeResult for more
+  /// information about the failure.
+  COREWEBVIEW2_UPDATE_RUNTIME_STATUS_FAILED,
+} COREWEBVIEW2_UPDATE_RUNTIME_STATUS;
+
+
+/// The TryUpdateRuntime operation result.
+[uuid(DD503E49-AB19-47C0-B2AD-6DDD09CC3E3A), object, pointer_default(unique)]
+interface ICoreWebView2ExperimentalUpdateRuntimeResult : IUnknown {
+
+  /// The status for the TryUpdateRuntime operation.
+  [propget] HRESULT Status(
+      [ out, retval ] COREWEBVIEW2_UPDATE_RUNTIME_STATUS * status);
+
+  /// The update error happened while trying to update Edge WebView2 Runtime.
+  [propget] HRESULT UpdateError([out, retval] HRESULT* updateError);
+}
+
+/// The caller implements this interface to receive the TryUpdateRuntime result.
+[uuid(F1D2D722-3721-499C-87F5-4C405260697A), object, pointer_default(unique)]
+interface ICoreWebView2ExperimentalTryUpdateRuntimeCompletedHandler : IUnknown {
+
+  /// Provides the result for the TryUpdateRuntime operation.
+  /// `errorCode` will be S_OK if the update operation can be performed
+  /// normally, regardless of whether we could update the Edge WebView2
+  /// Runtime. If an unexpected error interrupts the update operation, error
+  /// code of that unexpected error would be set as `errorCode`.
+  /// When update operation can be performed normally, but update resulted in
+  /// failure, like download failed, the error code would be presented as
+  /// `UpdateError` property of ICoreWebView2ExperimentalUpdateRuntimeResult.
+  HRESULT Invoke([in] HRESULT errorCode,
+                 [in] ICoreWebView2ExperimentalUpdateRuntimeResult * result);
+}
+
+/// This interface is an extension of the ICoreWebView2Environment. An object
+/// implementing the ICoreWebView2ExperimentalEnvironment3 interface will also
+/// implement ICoreWebView2Environment.
+[uuid(9A2BE885-7F0B-4B26-B6DD-C969BAA00BF1), object, pointer_default(unique)]
+interface ICoreWebView2ExperimentalEnvironment3 : IUnknown {
+  /// Try to update the installed Microsoft Edge WebView2 Runtime.
+  /// This will potentially result in a new version of the Edge WebView2
+  /// Runtime being installed and `NewBrowserVersionAvailable` event being raised.
+  /// There is no guarantee on the order of that event being raised and
+  /// TryUpdateRuntime's completed handler being invoked. Besides the
+  /// `NewBrowserVersionAvailable` event, there will be no impact to any
+  /// currently running WebView2s when the update is installed.
+  /// The latest version can always be queried using the
+  /// `GetAvailableCoreWebView2BrowserVersionString` API.
+  /// The TryUpdateRuntime method is only supported for an installed Edge WebView2
+  /// Runtime. When running a fixed version Edge WebView2 Runtime or non stable
+  /// channel Edge browser, this API will return `HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED)`.
+  /// There could only be one active TryUpdateRuntime operation, calling this API
```

Comment:
Is this limit per-environment, per process, or system-wide? If system-wide, how can the app know how many other active calls there are? Even if per-process, it may be hard for an app to know how many calls are active, because different components may be doing work independently.

Reply:
Updated the comment: 
To protect accidental abuse of the update service, the implementation throttles the calls of this API to 3 times within 5 minutes in a process.
Update runtime is not something multiple components should be calling independently. It should only be called by app's "updater". We don't want to encourage apps calling this API from multiple places.

Reply:
Channeling @david-risney: "(1) Throttling should be a status code instead of an error code, so apps can meaningfully back off when they hit the limit. (2) Should note that the throttling limit can change in the future."

Reply:
Added a new state as following, but still fail the API for multiple requests from the app:
  /// Edge WebView2 Runtime update is already running, which could be
  /// triggered by auto update or by other UpdateRuntime request from some app.
  COREWEBVIEW2_UPDATE_RUNTIME_STATUS_INSTALL_ALREADY_RUNNING,